harvester_e2e_tests/integration/test_vm_snapshot.py

```diff
@@ -0,0 +1,517 @@
+            vm_sh = vm_shell(user, pkey=privatekey)
+
+            deadline = datetime.now() + timedelta(seconds=wait_timeout)
+            while deadline > datetime.now():
+                try:
+                    vm_sh.connect(vm_ip, jumphost=h.client)
+                except ChannelException as e:
+                    print(e)
+                    sleep(3)
+                else:
+                    break
+            else:
+                raise AssertionError(f"Unable to login to {name}")
+
+            with vm_sh as sh:
+                action(sh)
+
+
+@pytest.mark.p0
+@pytest.mark.virtualmachines
+class TestVMSnapshot:
+    @pytest.mark.dependency(name="source_vm_snapshot")
+    def test_vm_snapshot_create(self, api_client,
+                                source_vm, vm_snapshot_name,
+                                host_shell, vm_shell,
+                                ssh_keypair, wait_timeout):
+        """
+        Test that the VM snapshot can be created.
+
+        Prerequisite:
+        A virtual machine has been created and is running.
+        """
+        name, ssh_user = source_vm
+
+        def action(sh):
+            _, _ = sh.exec_command("echo 123 > test.txt")
+            _, _ = sh.exec_command("sync")
+
+        vm_shell_do(name, api_client,
+                    host_shell, vm_shell,
+                    ssh_user, ssh_keypair,
+                    action, wait_timeout)
+
+        # Since `sync` isn't actually synchronous, wait a couple of
+        # seconds to let the I/O flush to disk.
+        sleep(2)
+
+        code, _ = api_client.vm_snapshots.create(name, vm_snapshot_name)
+        assert 201 == code
+
+        deadline = datetime.now() + timedelta(seconds=wait_timeout)
+        while deadline > datetime.now():
+            code, data = api_client.vm_snapshots.get(vm_snapshot_name)
+            if data.get("status", {}).get("readyToUse"):
+                break
+            print(f"waiting for {vm_snapshot_name} to be ready")
+            sleep(3)
+        else:
+            raise AssertionError(f"timed out waiting for {vm_snapshot_name} to be ready")
+
+        assert 200 == code
+        assert data.get("status", {}).get("readyToUse") is True
+
+    @pytest.mark.dependency(depends=["source_vm_snapshot"])
+    def test_restore_into_new_vm_from_vm_snapshot(self, api_client,
+                                                  restored_from_snapshot_vm, source_vm,
+                                                  ssh_keypair, host_shell,
+                                                  vm_shell, wait_timeout):
+        """
+        Test that restoring the `vm-snapshot` into a new virtual
+        machine results in a virtual machine with the expected
+        well-known file (`test.txt`) with the expected contents
+        (`123`).
+
+        Prerequisites:
+        1. The source VM from the first test case and its
+           snapshot (`vm-snapshot`).
+        2. A VM restored from `vm-snapshot`.
+        """
+
```

Comment:
I think this implement will not hit criteria, the VM created and took snapshot, but we didn't modify it after snapshot, so we can not guarantee the VM really restored from the snapshot or not.

and while restoring to new VM, we will not stop the source VM.

Reply:
🤔 I understand this logic for the test_replace_vm_with_vm_snapshot, but I'm not sure I follow for this test case. Should I add more assertions to the test proving that this is accessing a completely separate/new VM?

Reply:
In manual testing, we will test this as following steps:
1. Create the VM `vm1`
2. Write data into `vm1` (`vm1-state-1`)
3. Take snapshot for `vm1`
4. Write or modify exisiting data into `vm1` (`vm1-state-2`)
5. Restore snapshot into `vm1` or `vm1-new`
6. Data in restored VM should keep in `vm1-state1`, not `vm1-state-2`